cuda/include/vecmem/memory/memory_types.hpp

```diff
@@ -0,0 +1,18 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+namespace vecmem::memory {
+   enum class memory_type {
```

Comment:
I'm on board with having a common definition for the CUDA memory managers. But we should not make this global.

oneAPI's USM for instance calls "the 3rd kind" of memory "shared" memory. Not "managed" memory. And HIP does an entirely different thing with "registering host memory to be accessible from the device".

https://rocmdocs.amd.com/en/latest/ROCm_API_References/HIP_API/Memory-Management.html#hiphostregister

So the easiest would be to avoid handling all languages in one strict way.

I propose making this:

```c++
namespace vecmem::cuda {
   enum class memory_type {
      ...
   };
}
```

, and putting it in `vecmem/memory/cuda/memory_type.hpp`. (I.e. let's use the name of the enum as the file name...)

Reply:
So there's now, let's say, abstraction that could be used to define an enum of this kind?

Reply:
I'm not sure... I both want to stay away from over-generalising the code, limiting our options by that. At the same time, as I wrote in https://github.com/acts-project/vecmem/pull/5#pullrequestreview-586842764, a single definition would make things a little easier.

I'm still pondering what we should be doing...

Reply:
I'm pretty conflicted about this to be honest, same as with #5 I worry that if we don't make some steps to generalize things we are going to end up with an unmaintainable amount of boilerplate. Already with the SYCL implementation I think there is a lot of stuff there that could be de-duplicated and abstracted away in a way very similar to the CUDA implementation.